include/nbl/builtin/hlsl/workgroup/fft.hlsl

```diff
@@ -84,20 +91,34 @@ struct FFT<2,false, Scalar, device_capabilities>
         complex_t<Scalar> lo = {loVec.x, loVec.y};  
         complex_t<Scalar> hi = {hiVec.x, hiVec.y};
 
-        // special first iteration
-        if (_NBL_HLSL_WORKGROUP_SIZE_ > glsl::gl_SubgroupSize())
+        // special first iteration - only if workgroupsize > subgroupsize
+        if (_NBL_HLSL_WORKGROUP_SIZE_ ^ glsl::gl_SubgroupSize())
             fft::DIF<Scalar>::radix2(fft::twiddle<false, Scalar>(threadID, _NBL_HLSL_WORKGROUP_SIZE_), lo, hi); 
 
         // Run bigger steps until Subgroup-sized
         for (uint32_t stride = _NBL_HLSL_WORKGROUP_SIZE_ >> 1; stride > glsl::gl_SubgroupSize(); stride >>= 1)
+        {   
+            // If at least one loop was executed, we must wait for all threads to get their values before we write to shared mem again
+            if ( !(stride & (_NBL_HLSL_WORKGROUP_SIZE_ >> 1)) )
+                sharedmemAdaptor.workgroupExecutionAndMemoryBarrier(); 
             FFT_loop<SharedMemoryAccessor>(stride, lo, hi, threadID, sharedmemAdaptor);
-  
+        }
+
+        // special last workgroup-shuffle - only if workgroupsize > subgroupsize
+        if (_NBL_HLSL_WORKGROUP_SIZE_ ^ glsl::gl_SubgroupSize()) 
+        {
+            // Wait for all threads to be done with reads in the last loop before writing to shared mem      
+            sharedmemAdaptor.workgroupExecutionAndMemoryBarrier(); 
```

Comment:
if you barrier after like in https://github.com/Devsh-Graphics-Programming/Nabla/pull/698/files/dcd78d64fbdda32f25efeea29d5320d542c34482..74b978af5275b1b6d9b6a4fe9a39c023ebfb1899#r1655572151

then you don't need a barrier here